src/go/k8s/internal/client/factory_test.go

```diff
@@ -0,0 +1,242 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/redpanda-data/helm-charts/pkg/helm"
+	"github.com/redpanda-data/helm-charts/pkg/kube"
+	"github.com/redpanda-data/helm-charts/pkg/testutil"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/redpanda-data/redpanda-operator/src/go/k8s/pkg/k3d"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kadm"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var chartVersion = ""
+
+func ensureMapAndSetValue(values map[string]any, name, key string, value any) {
```

Comment:
You could also use the unstructured package from Kubernetes to do this.

Reply:
Ah, I'll take a look to see what map helpers `unstructured` has in it, thanks for the tip!

Reply:
😞 just tried the `unstructured` package, it attempts to use `DeepCopy` on whatever value you pass it as an interface, which panics when you're trying to set primitives, so think I need to keep this helper unless you know of some other package out there I can use to "deep set" map keys pretty easily.

Reply:
Mind sharing the invocation you used? I see a call to [`DeepCopyJSONValue`](https://github.com/kubernetes/apimachinery/blob/a8f449e276fe566efddb149992049c78f0088492/pkg/apis/meta/v1/unstructured/helpers.go#L208-L230) but that looks like it handles primitives well and I've not encountered panics in [my usages](https://github.com/redpanda-data/redpanda-operator/blob/acc04250c2a14225f885b3c49926036718e965a9/src/go/k8s/api/redpanda/v1alpha2/redpanda_types_test.go#L191-L196) 🤔 .

Reply:
Yep, so that's what I tried:

```
require.NoError(t, unstructured.SetNestedField(values, 1, "statefulset", "replicas"))
```

Here's the stack trace:

```
panic: cannot deep copy int [recovered]
	panic: cannot deep copy int
...
in-arm64/src/runtime/panic.go:770 +0x124
k8s.io/apimachinery/pkg/runtime.DeepCopyJSONValue({0x106ae9aa0?, 0x10693d010?})
	/Users/andrewstucki/go/pkg/mod/k8s.io/apimachinery@v0.29.5/pkg/runtime/converter.go:639 +0x24c
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.SetNestedField(0x1400062f560, {0x106ae9aa0?, 0x10693d010?}, {0x14000671e70, 0x2, 0x2})
...
```

Reply:
Fun enough this little go program:

```
package main

import (
	"fmt"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

func main() {
	d := map[string]any{}
	unstructured.SetNestedField(d, "A", "a", "b")
	fmt.Println("set string")
	unstructured.SetNestedField(d, 1.0, "a", "c")
	fmt.Println("set float")
	unstructured.SetNestedField(d, false, "a", "d")
	fmt.Println("set bool")
	unstructured.SetNestedField(d, 1, "a", "e")
	fmt.Println("set int")
}
```

Seems to say that other primitives are completely fine:

```
~> go run main.go
set string
set float
set bool
panic: cannot deep copy int
```

Reply:
That's fun, looks like if I use an `int64` instead it'll work, feels odd that they decided on this subset of primitive types:

https://github.com/kubernetes/apimachinery/blob/a8a2284d318b9213fca20b0058e969941ce1c90b/pkg/runtime/converter.go#L636